packages/apollo-server-codegen/src/Translators/TypeScriptTranslator.ts

```diff
@@ -0,0 +1,364 @@
+    );
+
+    return [
+      header,
+      resolvers,
+      ...[...objects, ...enums, ...scalars, ...unions, ...inputObjects].map(
+        def => def.translate(this)
+      )
+    ].join("\n");
+  }
+
+  private generateHeader() {
+    // Generate some utility higher order types that we'll reference later on.
+    return `// This is a machine generated file.
+// Use "apollo service:codegen" to regenerate.
+type PromiseOrValue<T> = Promise<T> | T
+type Nullable<T> = T | null | undefined
+type Index<Map extends Record<string, any>, Key extends string> = Map[Key] extends object ? Map[Key] : unknown
+`;
+  }
+
+  private generateTopLevelResolvers(
+    types: string[],
+    enums: string[],
+    scalars: string[]
+  ) {
+    return [
+      `export interface Resolvers<TContext = {}, TInternalReps = {}> {`,
+      ...types.map(type =>
+        this.rootTypes.includes(type)
+          ? `${type}: ${type}Resolver<TContext, TInternalReps>`
+          : `${type}?: ${type}Resolver<TContext, TInternalReps>`
+      ),
+      ...scalars.map(scalar => `  ${scalar}: any`),
+      ...(this.options.__experimentalInternalEnumValueSupport
+        ? enums.map(
+            enumName => `${enumName}: { [external: ${enumName}External]: any }`
+          )
+        : []),
+      `}\n`
+    ].join("\n");
+  }
+
+  public translateDescription(t: IR.Description): string {
+    return "/**\n * " + t.description.split("\n").join("\n * ") + "\n */\n";
+  }
+
+  public translateNamedType(t: IR.NamedType, nullable: boolean): string {
+    const getName = () => {
+      switch (t.name) {
+        case "Int":
+          return "number";
+        case "Boolean":
+          return "boolean";
+        case "ID":
+          return "string";
+        case "String":
+          return "string";
+        default:
+          return t.name;
+      }
+    };
+
+    return nullable ? `Nullable<${getName()}>` : getName();
+  }
+
+  public translateNonNullType(t: IR.NonNullType): string {
+    return t.base.translate(this, false);
+  }
+
+  public translateListType(t: IR.ListType, nullable: boolean): string {
+    return nullable
+      ? `Nullable<Array<${t.base.translate(this)}>>`
+      : `Array<${t.base.translate(this)}>`;
+  }
+
+  public translateCompoundType(t: IR.CompoundType): string {
+    return [
+      "{ ",
+      ...t.types.map(type => `${type.name}: ${type.type.translate(this)}, `),
+      "}"
+    ].join("");
+  }
+
+  public translateArgumentDefinition(t: IR.ArgumentDefinition): string {
+    return [
+      t.description.translate(this),
+      t.name,
+      // Bit of hack here to use optional types instead of Nullable, for better compatibility with
+      // destructuring default assignment: ({arg = default}) => ...
+      t.type instanceof IR.NonNullType ? `: ` : `?: `,
+      new IR.NonNullType(t.type).translate(this)
+    ].join("");
+  }
+
+  public translateResolverDefinition(t: IR.ResolverDefinition): string {
+    if (t.isNotProvidedAndExternal) {
```

Comment:
This is mentioned above, but I don't think this is the correct behavior

Reply:
Using keys and provides now